# **Neural Network Cardiotocographic**

# **Pre-Processing of the data**

In [1]:
# Cardio dataset
cardio = read.csv("../input/cardiotocographic/Cardiotocographic.csv")
# First 6 rows of the dataset
head(cardio)

LB  AC          FM UC          DL          DS DP          ASTV MSTV ALTV ⋯
1 120 0.000000000 0  0.000000000 0.000000000 0  0.000000000 73   0.5  43   ⋯
2 132 0.006379585 0  0.006379585 0.003189793 0  0.000000000 17   2.1   0   ⋯
3 133 0.003322259 0  0.008305648 0.003322259 0  0.000000000 16   2.1   0   ⋯
4 134 0.002560819 0  0.007682458 0.002560819 0  0.000000000 16   2.4   0   ⋯
5 132 0.006514658 0  0.008143322 0.000000000 0  0.000000000 16   2.4   0   ⋯
6 134 0.001049318 0  0.010493179 0.009443861 0  0.002098636 26   5.9   0   ⋯
  Min Max Nmax Nzeros Mode Mean Median Variance Tendency NSP
1 62  126  2   0      120  137  121     73      1        2  
2 68  198  6   1      141  136  140     12      0        1  
3 68  198  5   1      141  135  138     13      0        1  
4 53  170 11   0      137  134  137     13      1        1  
5 53  170  9   0      137  136  138     11      1        1  
6 50  200  5   3       76  107  107    170      0        3

In [2]:
# Stucture of the dataset
str(cardio)

'data.frame':	2126 obs. of  22 variables:
 $ LB      : int  120 132 133 134 132 134 134 122 122 122 ...
 $ AC      : num  0 0.00638 0.00332 0.00256 0.00651 ...
 $ FM      : num  0 0 0 0 0 0 0 0 0 0 ...
 $ UC      : num  0 0.00638 0.00831 0.00768 0.00814 ...
 $ DL      : num  0 0.00319 0.00332 0.00256 0 ...
 $ DS      : num  0 0 0 0 0 0 0 0 0 0 ...
 $ DP      : num  0 0 0 0 0 ...
 $ ASTV    : int  73 17 16 16 16 26 29 83 84 86 ...
 $ MSTV    : num  0.5 2.1 2.1 2.4 2.4 5.9 6.3 0.5 0.5 0.3 ...
 $ ALTV    : int  43 0 0 0 0 0 0 6 5 6 ...
 $ MLTV    : num  2.4 10.4 13.4 23 19.9 0 0 15.6 13.6 10.6 ...
 $ Width   : int  64 130 130 117 117 150 150 68 68 68 ...
 $ Min     : int  62 68 68 53 53 50 50 62 62 62 ...
 $ Max     : int  126 198 198 170 170 200 200 130 130 130 ...
 $ Nmax    : int  2 6 5 11 9 5 6 0 0 1 ...
 $ Nzeros  : int  0 1 1 0 0 3 3 0 0 0 ...
 $ Mode    : int  120 141 141 137 137 76 71 122 122 122 ...
 $ Mean    : int  137 136 135 134 136 107 107 122 122 122 ...
 $ Median  : int  1

In [3]:
# Summary of the dataset
summary(cardio)

       LB              AC                 FM                 UC          
 Min.   :106.0   Min.   :0.000000   Min.   :0.000000   Min.   :0.000000  
 1st Qu.:126.0   1st Qu.:0.000000   1st Qu.:0.000000   1st Qu.:0.001876  
 Median :133.0   Median :0.001630   Median :0.000000   Median :0.004482  
 Mean   :133.3   Mean   :0.003170   Mean   :0.009474   Mean   :0.004357  
 3rd Qu.:140.0   3rd Qu.:0.005631   3rd Qu.:0.002512   3rd Qu.:0.006525  
 Max.   :160.0   Max.   :0.019284   Max.   :0.480634   Max.   :0.014925  
       DL                 DS                  DP                 ASTV      
 Min.   :0.000000   Min.   :0.000e+00   Min.   :0.0000000   Min.   :12.00  
 1st Qu.:0.000000   1st Qu.:0.000e+00   1st Qu.:0.0000000   1st Qu.:32.00  
 Median :0.000000   Median :0.000e+00   Median :0.0000000   Median :49.00  
 Mean   :0.001885   Mean   :3.585e-06   Mean   :0.0001566   Mean   :46.99  
 3rd Qu.:0.003264   3rd Qu.:0.000e+00   3rd Qu.:0.0000000   3rd Qu.:61.00  
 Max.   :0.015385   Max.  

*No NA values are there in the dataset. But we need to convert the variables to same scale by normalizing all the variables*

# **Normalization of the variables**

In [4]:
cardio$LB = (cardio$LB-min(cardio$LB))/(max(cardio$LB)-min(cardio$LB))

cardio$AC = (cardio$AC-min(cardio$AC))/(max(cardio$AC)-min(cardio$AC))

cardio$FM = (cardio$FM-min(cardio$FM))/(max(cardio$FM)-min(cardio$FM))

cardio$UC = (cardio$UC-min(cardio$UC))/(max(cardio$UC)-min(cardio$UC))

cardio$DL = (cardio$DL-min(cardio$DL))/(max(cardio$DL)-min(cardio$DL))

cardio$DS = (cardio$DS-min(cardio$DS))/(max(cardio$DS)-min(cardio$DS))

cardio$DP = (cardio$DP-min(cardio$DP))/(max(cardio$DP)-min(cardio$DP))

cardio$ASTV = (cardio$ASTV-min(cardio$ASTV))/(max(cardio$ASTV)-min(cardio$ASTV))

cardio$MSTV = (cardio$MSTV-min(cardio$MSTV))/(max(cardio$MSTV)-min(cardio$MSTV))

cardio$ALTV = (cardio$ALTV-min(cardio$ALTV))/(max(cardio$ALTV)-min(cardio$ALTV))

cardio$MLTV = (cardio$MLTV-min(cardio$MLTV))/(max(cardio$MLTV)-min(cardio$MLTV))

cardio$Width = (cardio$Width-min(cardio$Width))/(max(cardio$Width)-min(cardio$Width))

cardio$Min = (cardio$Min-min(cardio$Min))/(max(cardio$Min)-min(cardio$Min))

cardio$Max = (cardio$Max-min(cardio$Max))/(max(cardio$Max)-min(cardio$Max))

cardio$Nmax = (cardio$Nmax-min(cardio$Nmax))/(max(cardio$Nmax)-min(cardio$Nmax))

cardio$Nzeros = (cardio$Nzeros-min(cardio$Nzeros))/(max(cardio$Nzeros)-min(cardio$Nzeros))

cardio$Mode = (cardio$Mode-min(cardio$Mode))/(max(cardio$Mode)-min(cardio$Mode))

cardio$Mean = (cardio$Mean-min(cardio$Mean))/(max(cardio$Mean)-min(cardio$Mean))

cardio$Median = (cardio$Median-min(cardio$Median))/(max(cardio$Median)-min(cardio$Median))

cardio$Variance = (cardio$Variance-min(cardio$Variance))/(max(cardio$Variance)-min(cardio$Variance))

cardio$Tendency = (cardio$Tendency-min(cardio$Tendency))/(max(cardio$Tendency)-min(cardio$Tendency))


In [5]:
# Structure of the dataset
str(cardio)

'data.frame':	2126 obs. of  22 variables:
 $ LB      : num  0.259 0.481 0.5 0.519 0.481 ...
 $ AC      : num  0 0.331 0.172 0.133 0.338 ...
 $ FM      : num  0 0 0 0 0 0 0 0 0 0 ...
 $ UC      : num  0 0.427 0.556 0.515 0.546 ...
 $ DL      : num  0 0.207 0.216 0.166 0 ...
 $ DS      : num  0 0 0 0 0 0 0 0 0 0 ...
 $ DP      : num  0 0 0 0 0 ...
 $ ASTV    : num  0.8133 0.0667 0.0533 0.0533 0.0533 ...
 $ MSTV    : num  0.0441 0.2794 0.2794 0.3235 0.3235 ...
 $ ALTV    : num  0.473 0 0 0 0 ...
 $ MLTV    : num  0.0473 0.2051 0.2643 0.4536 0.3925 ...
 $ Width   : num  0.345 0.718 0.718 0.644 0.644 ...
 $ Min     : num  0.1101 0.1651 0.1651 0.0275 0.0275 ...
 $ Max     : num  0.0345 0.6552 0.6552 0.4138 0.4138 ...
 $ Nmax    : num  0.111 0.333 0.278 0.611 0.5 ...
 $ Nzeros  : num  0 0.1 0.1 0 0 0.3 0.3 0 0 0 ...
 $ Mode    : num  0.472 0.638 0.638 0.606 0.606 ...
 $ Mean    : num  0.587 0.578 0.569 0.56 0.578 ...
 $ Median  : num  0.404 0.578 0.56 0.55 0.56 ...
 $ Variance: num  0.2714 0.

# **Data Partition**

In [6]:
# Setting seed 
set.seed(1234)

# Dividing the dataset into train and test dataset
pd = sample(2,nrow(cardio),replace= TRUE,prob=c(0.8,0.2))

# Train dataset
train = cardio[pd==1,] 

# No. of record in the train dataset from cardio dataset
nrow(train)/nrow(cardio)*100

# Test dataset
test = cardio[pd==2,]

# No. of record in the test dataset from cardio dataset
nrow(test)/nrow(cardio)*100

[1] 80.80903

[1] 19.19097

*We have 80.81% and 19.19% of the records in the train and test dataset respectively from the cardio dataset.*

# **Building Neural Network**

In [7]:
# Importing library neuralnet
library(neuralnet)

# Neural Network
nn = neuralnet(NSP ~ ., data = train)

# **Prediction**

In [8]:
# Predicting neural net from test dataset
pred_nn = compute(nn, test[,-22])

# Assigning the result to p_nn
p_nn = pred_nn$net.result

# Prediction
head(p_nn, 5)

# Actual values
head(test[,22], 5)

[,1]    
5  1.013863
14 1.014168
16 1.113719
26 2.800365
28 2.820696

[1] 1 1 1 3 3

In [9]:
# Prediction vs Actual Value
cbind(head(p_nn, 15), head(test[,22],15))

[,1]     [,2]
5  1.013863 1   
14 1.014168 1   
16 1.113719 1   
26 2.800365 3   
28 2.820696 3   
29 1.757048 2   
39 1.015472 1   
40 1.014101 1   
60 1.014156 1   
61 1.017477 1   
72 1.017679 1   
81 1.074106 1   
86 1.013787 1   
90 1.180691 2   
92 2.445551 2

*The prediction for 5th record is 1.013863 which is close to 1, and this is a accurate prediction. Similarly the prediction is accurat for all other records shown above. Except for 90th we the prediction is close to 2 (1.180691), but the actual value is 2.*


In [10]:
# Prediction vs Actual Value
cbind(tail(p_nn, 15), tail(test[,22],15))

[,1]     [,2]
2046 1.049020 1   
2053 1.076503 1   
2056 1.472116 1   
2061 1.740470 1   
2066 1.534360 1   
2073 1.033262 1   
2076 1.041825 1   
2089 1.990645 1   
2091 1.791536 1   
2098 1.301036 1   
2102 1.056980 1   
2104 1.100274 1   
2109 1.125183 1   
2116 1.784735 1   
2121 1.418997 1

*We can see some **wrong** prediction in the above table. Like for 2061th the value is 1.740470 which is close to 2 but the actual calue is 1. Similary the prediction is wrong in case of 2066th, 2089th, 2091th and 2116th.*

# **Accuracy Level**

In [11]:
# Predicting the value as categories 1, 2 or 3
# Assign 3 if the value is more than 2.5, if more than 1.5, otherwise 1.
cm = ifelse(p_nn > 2.5, 3, ifelse(p_nn > 1.5, 2, 1))

# Confusion Matrix
tab_nn=table(cm,test$NSP)

# Printing Confusion Matrix
print(tab_nn)

   
cm    1   2   3
  1 293  31   2
  2  14  28   4
  3   0  10  26


In [12]:
# Accuracy Percentage
sum(diag(tab_nn))/sum(tab_nn)*100

[1] 85.04902

*The network is giving 85.05% accuracy. For this high value we can consider the network, and the model is likely to give correct prediction 85.05% of the time.*